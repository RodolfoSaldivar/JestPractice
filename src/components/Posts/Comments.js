import React from 'react';
import { connect } from 'react-redux';
import Spinner from '../General/Spinner';
import Fatal from '../General/Fatal';

export const Comments = (props) => {
	if (props.com_error) {
		return <Fatal message={props.com_error} />;
	}
	if (props.com_loading && !props.comments.length) {
		return <Spinner />;
	}

	const displayComments = () =>
		props.comments.map((comment) => (
			<li key={comment.id}>
				<b>
					<u>{comment.email}</u>
				</b>
				<br />
				{comment.body}
			</li>
		));

	return <ul>{displayComments()}</ul>;
};

const mapStateToProps = ({ postsReducer }) => postsReducer;

export default connect(mapStateToProps)(Comments);
