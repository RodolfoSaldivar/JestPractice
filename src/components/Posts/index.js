import React, { Component } from 'react';
import { connect } from 'react-redux';
import { withStyles } from '@material-ui/core/styles';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';

import Spinner from '../General/Spinner';
import Fatal from '../General/Fatal';
import Comments from './Comments';

import * as usersActions from '../../actions/usersActions';
import * as postsActions from '../../actions/postsActions';

const { getAll: usersGetAll } = usersActions;
const { getByUser: postsGetByUser, openClose, getComments } = postsActions;

//================================================

const styles = (theme) => ({
	root: {
		width: '100%'
	}
});

//================================================

export class Posts extends Component {
	async componentDidMount() {
		const {
			usersGetAll,
			match: {
				params: { key }
			},
			postsGetByUser
		} = this.props;

		if (!this.props.usersReducer.users.length) {
			await usersGetAll();
		}
		if (this.props.usersReducer.error) {
			return;
		}
		if (!('posts_key' in this.props.usersReducer.users[key])) {
			await postsGetByUser(key);
		}
	}

	displayUser = () => {
		const {
			match: {
				params: { key }
			},
			usersReducer
		} = this.props;

		if (usersReducer.error) {
			return <Fatal message={usersReducer.error} />;
		}
		if (!usersReducer.users.length || usersReducer.loading) {
			return <Spinner />;
		}

		const name = usersReducer.users[key].name;

		return <h1>Posts of: {name}</h1>;
	};

	displayPosts = () => {
		const {
			usersReducer,
			usersReducer: { users },
			postsReducer,
			postsReducer: { posts },
			match: {
				params: { key }
			}
		} = this.props;

		if (!users.length) return;
		if (usersReducer.error) return;
		if (postsReducer.loading) {
			return <Spinner />;
		}
		if (postsReducer.error) {
			return <Fatal message={postsReducer.error} />;
		}
		if (!posts.length) return;
		if (!('posts_key' in users[key])) return;

		const { posts_key } = users[key];
		return this.showInfo(posts[posts_key], posts_key);
	};

	showInfo = (posts, post_key) =>
		posts.map((post, com_key) => (
			<ExpansionPanel
				key={post.id}
				onClick={() => this.showComments(post_key, com_key, post.comments)}
				elevation={3}
			>
				<ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
					<h2>{post.title}</h2>
				</ExpansionPanelSummary>
				<ExpansionPanelDetails>
					<div>
						<h3>{post.body}</h3>
						{post.open && <Comments comments={post.comments} />}
					</div>
				</ExpansionPanelDetails>
			</ExpansionPanel>
		));

	showComments = (post_key, com_key, comments) => {
		this.props.openClose(post_key, com_key);
		if (!comments.length) {
			this.props.getComments(post_key, com_key);
		}
	};

	render() {
		const { classes } = this.props;
		return (
			<div>
				{this.displayUser()}
				<div className={classes.root}>{this.displayPosts()}</div>
			</div>
		);
	}
}

const mapStateToProps = ({ usersReducer, postsReducer }) => {
	return { usersReducer, postsReducer };
};

const mapDispatchToProps = {
	usersGetAll,
	postsGetByUser,
	openClose,
	getComments
};

export default withStyles(styles)(
	connect(
		mapStateToProps,
		mapDispatchToProps
	)(Posts)
);
