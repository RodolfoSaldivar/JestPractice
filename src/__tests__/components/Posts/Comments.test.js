import React from 'react';
import { shallow } from 'enzyme';
import fixtures from '../../fixtures/posts';
import { Comments } from '../../../components/Posts/Comments';

const PROPS = {
	...fixtures.initial_state,
	comments: []
};

describe('<Comments />', () => {
	it('Renders correctly', () => {
		const comments = shallow(<Comments {...PROPS} />);
		expect(comments).toMatchSnapshot();
	});

	it('Returns <Fatal />', () => {
		const comments = shallow(<Comments {...PROPS} com_error="something" />);
		expect(comments).toMatchSnapshot();
		expect(comments.find('Fatal')).toHaveLength(1);
	});

	it('Returns <Spinner />', () => {
		const comments = shallow(<Comments {...PROPS} com_loading />);
		expect(comments).toMatchSnapshot();
		expect(comments.find('Spinner')).toHaveLength(1);
	});

	it('Returns "ul" with "li"', () => {
		const comments = shallow(
			<Comments {...PROPS} comments={fixtures.comments} />
		);
		expect(comments).toMatchSnapshot();
		expect(comments.find('ul')).toHaveLength(1);
		expect(comments.find('li').exists()).toEqual(true);
	});
});
