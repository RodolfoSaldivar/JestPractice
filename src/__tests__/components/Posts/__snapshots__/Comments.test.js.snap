// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Comments /> Renders correctly 1`] = `<ul />`;

exports[`<Comments /> Returns "ul" with "li" 1`] = `
<ul>
  <li
    key="1"
  >
    <b>
      <u>
        Eliseo@gardner.biz
      </u>
    </b>
    <br />
    laudantium enim quasi est quidem magnam voluptate ipsam eos tempora quo necessitatibus dolor quam autem quasi reiciendis et nam sapiente accusantium
  </li>
  <li
    key="2"
  >
    <b>
      <u>
        Jayne_Kuhic@sydney.com
      </u>
    </b>
    <br />
    est natus enim nihil est dolore omnis voluptatem numquam et omnis occaecati quod ullam at voluptatem error expedita pariatur nihil sint nostrum voluptatem reiciendis et
  </li>
</ul>
`;

exports[`<Comments /> Returns <Fatal /> 1`] = `
<Fatal
  message="something"
/>
`;

exports[`<Comments /> Returns <Spinner /> 1`] = `<Spinner />`;
